### PR TITLE
Add Nucleo-64 stm32f303RE board

### PR DESCRIPTION
**Pull request by paulbendixen**
Added SoC and board for another nucleo board

### CODE REVIEW

File: boards/arm/nucleo_f303re/Kconfig.defconfig
Comment (henrikbrixandersen) [inline boards/arm/nucleo_f303re/Kconfig.defconfig]: Drop the extra linefeed here.

File: boards/arm/nucleo_f303re/Kconfig.defconfig
Comment (henrikbrixandersen) [inline boards/arm/nucleo_f303re/Kconfig.defconfig]: This should be `BOARD_NUCLEO_F303RE`.

File: boards/arm/nucleo_f303re/doc/index.rst
Comment (henrikbrixandersen) [inline boards/arm/nucleo_f303re/doc/index.rst]: This should be `_nucleo_f303re_board`.

File: boards/arm/nucleo_f303re/doc/index.rst
Comment (henrikbrixandersen) [inline boards/arm/nucleo_f303re/doc/index.rst]: This should be referring to the `nucleo_f303re` directory.

File: boards/arm/nucleo_f303re/doc/index.rst
Comment (henrikbrixandersen) [inline boards/arm/nucleo_f303re/doc/index.rst]: Extra linefeed here.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
@paulbendixen Thank you for your first contribution to Zephyr! A few minor changes suggested here.

File: soc/arm/st_stm32/stm32f3/Kconfig.defconfig.stm32f303xe
Comment (erwango) [inline soc/arm/st_stm32/stm32f3/Kconfig.defconfig.stm32f303xe]: Already enabled in soc/arm/st_stm32/common/Kconfig.defconfig.series

File: soc/arm/st_stm32/stm32f3/Kconfig.defconfig.stm32f303xe
Comment (erwango) [inline soc/arm/st_stm32/stm32f3/Kconfig.defconfig.stm32f303xe]: Already enabled in soc/arm/st_stm32/stm32f3/Kconfig.defconfig.series

File: boards/arm/nucleo_f303re/nucleo_f303re.dts
Comment (erwango) [inline boards/arm/nucleo_f303re/nucleo_f303re.dts]: Break a line between nodes

File: boards/arm/nucleo_f303re/nucleo_f303re.dts
Comment (erwango) [inline boards/arm/nucleo_f303re/nucleo_f303re.dts]: usart3 pin are not configured.

File: boards/arm/nucleo_f303re/nucleo_f303re.dts
Comment (erwango) [inline boards/arm/nucleo_f303re/nucleo_f303re.dts]: usart2 and uart3 are not enabled in Kconfig

**Review by erwango — CHANGES_REQUESTED** [review]

File: boards/arm/nucleo_f303re/nucleo_f303re.dts
Comment (paulbendixen) [reply boards/arm/nucleo_f303re/nucleo_f303re.dts]: No, but `usart2` is set to be the configured output, `UART_2` is also set to default enabled in Kconfig.defconfig.
in `pinmux.c` the pins configured for `#ifdef CONFIG_UART_2` are `USART_2` pins.
Should there be some sort of renaming?

Comment (erwango) [reply boards/arm/nucleo_f303re/nucleo_f303re.dts]: You can have a look to other board to see how it is handled.
My request is to have a coherent configuration across dts/pinmux/kconfig..

**Comment by erwango** [comment]
@paulbendixen, thanks for taking into account the reported comments, but please fix them directly in faulty commits, we don't merge fix commits.

File: boards/arm/nucleo_f303re/nucleo_f303re.dts
Comment (paulbendixen) [reply boards/arm/nucleo_f303re/nucleo_f303re.dts]: ok looking into this

File: soc/arm/st_stm32/stm32f3/Kconfig.defconfig.stm32f303xe
Comment (erwango) [inline soc/arm/st_stm32/stm32f3/Kconfig.defconfig.stm32f303xe]: Ports A and B are enabled in soc/arm/st_stm32/common/Kconfig.defconfig.series

File: soc/arm/st_stm32/stm32f3/Kconfig.defconfig.stm32f303xe
Comment (erwango) [inline soc/arm/st_stm32/stm32f3/Kconfig.defconfig.stm32f303xe]: 85

File: boards/arm/nucleo_f303re/pinmux.c
Comment (erwango) [inline boards/arm/nucleo_f303re/pinmux.c]: Have you been able to test all of these ?

**Review by erwango — CHANGES_REQUESTED** [review]
Please check out zephyrbot advices (signed-off-by and such).

Also, please check that Kconfig + dts + pinmux are in sync.
I see there are inconsistencies in some boards in tree, I'll fix those.

File: soc/arm/st_stm32/stm32f3/Kconfig.defconfig.stm32f303xe
Comment (paulbendixen) [reply soc/arm/st_stm32/stm32f3/Kconfig.defconfig.stm32f303xe]: Shouldn't it be 84 as that is the highest number?
Just following https://github.com/zephyrproject-rtos/zephyr/commit/ae2742dc6bf3208a308dd047bec3d9c127492966
Couldn't find better documentation for the value

Comment (erwango) [reply soc/arm/st_stm32/stm32f3/Kconfig.defconfig.stm32f303xe]: Well, number should actually be highest IRQ number +1 as is starts from 0.

Comment (paulbendixen) [reply soc/arm/st_stm32/stm32f3/Kconfig.defconfig.stm32f303xe]: Ok

**Comment by paulbendixen** [comment]
I haven't been able to figure out how you want the drivers named

File: boards/arm/nucleo_f303re/arduino_r3_connector.dtsi
Comment (erwango) [inline boards/arm/nucleo_f303re/arduino_r3_connector.dtsi]: i2c not enabled in board.dts and in pinmux file, please remove this line.

File: boards/arm/nucleo_f303re/arduino_r3_connector.dtsi
Comment (erwango) [inline boards/arm/nucleo_f303re/arduino_r3_connector.dtsi]: spi not enabled in board.dts and in pinmux file, please remove this line.

File: boards/arm/nucleo_f303re/doc/index.rst
Comment (erwango) [inline boards/arm/nucleo_f303re/doc/index.rst]: i2c not enabled in board.dts and in pinmux file, please remove this line.

File: boards/arm/nucleo_f303re/doc/index.rst
Comment (erwango) [inline boards/arm/nucleo_f303re/doc/index.rst]: pwm not enabled in board.dts and in pinmux file, please remove this line.

File: boards/arm/nucleo_f303re/doc/index.rst
Comment (erwango) [inline boards/arm/nucleo_f303re/doc/index.rst]: spi not enabled in board.dts and in pinmux file, please remove this line.

File: boards/arm/nucleo_f303re/doc/index.rst
Comment (erwango) [inline boards/arm/nucleo_f303re/doc/index.rst]: i2c, spi, pwm not enabled in board.dts and in pinmux file, please remove this line.

File: boards/arm/nucleo_f303re/nucleo_f303re.yaml
Comment (erwango) [inline boards/arm/nucleo_f303re/nucleo_f303re.yaml]: `- counter`

(since you enabled rtc in board.dts)

File: boards/arm/nucleo_f303re/doc/index.rst
Comment (erwango) [inline boards/arm/nucleo_f303re/doc/index.rst]: You can add RTC here since enabled in board.dts file

**Review by erwango — CHANGES_REQUESTED** [review]
Few remaining comments to align board dts, pinmux doc and yaml.
I've taken dts as the reference.
Otherwise this is fine

**Review by erwango — APPROVED** [review]
Approved.
@paulbendixen You could add additional peripherals (spi, i2c, ..) on this board in a new PR (one peripheral per commit would be nice) if needed.

**Review by henrikbrixandersen — APPROVED** [review]

File: boards/arm/nucleo_f303re/nucleo_f303re.dts
Comment (henrikbrixandersen) [inline boards/arm/nucleo_f303re/nucleo_f303re.dts]: @paulbendixen You need to rebase on master and fix the GPIO device tree flags.

File: boards/arm/nucleo_f303re/nucleo_f303re.dts
Comment (henrikbrixandersen) [inline boards/arm/nucleo_f303re/nucleo_f303re.dts]: Same here.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: boards/arm/nucleo_f303re/nucleo_f303re.dts
Comment (erwango) [reply boards/arm/nucleo_f303re/nucleo_f303re.dts]: Thanks @henrikbrixandersen !

**Comment by galak** [comment]
Please look to run the doc images through something like https://tinypng.com/ to reduce size.

**Comment by paulbendixen** [comment]
All images are < 250KB in size. I don't think that's the problem.

**Review by henrikbrixandersen — APPROVED** [review]

**Review by erwango — APPROVED** [review]

File: boards/arm/nucleo_f303re/nucleo_f303re_defconfig
Comment (erwango) [inline boards/arm/nucleo_f303re/nucleo_f303re_defconfig]: This symbol is now (logically) deduced from SOC symbols. This is a recent change and assigning it in board's file generates an error at build (as we could see in shippable reports)
Please remove.

**Review by erwango — CHANGES_REQUESTED** [review]

File: boards/arm/nucleo_f303re/nucleo_f303re_defconfig
Comment (paulbendixen) [reply boards/arm/nucleo_f303re/nucleo_f303re_defconfig]: K

**Review by erwango — APPROVED** [review]

**Comment by jhedberg** [comment]
This is tagged for 2.2 but we're already in the stabilisation phase and it's not a bug fix. Do you want to bring this to the TSC to request an exception to be made, or should the target milestone be changed?

**Comment by erwango** [comment]
> This is tagged for 2.2 but we're already in the stabilisation phase and it's not a bug fix. Do you want to bring this to the TSC to request an exception to be made, or should the target milestone be changed?

@jhedberg this PR was tagged 2.2 and ready to be merged before closing merge window. It has missed a maintainer to pass-by and merge it on time. Though, on my side there is no specific need and that's ok to shift to 2.3. 
I let @paulbendixen say if it's fine on his side.

File: boards/arm/nucleo_f303re/Kconfig.defconfig
Comment (erwango) [inline boards/arm/nucleo_f303re/Kconfig.defconfig]: Please apply  recently merged https://github.com/zephyrproject-rtos/zephyr/pull/22666.

**Pull request merged**
